hedera-node/hedera-app-spi/src/main/java/com/hedera/node/app/spi/state/WritableKVStateBase.java

```diff
@@ -28,15 +28,22 @@
  */
 public abstract class WritableKVStateBase<K, V> extends ReadableKVStateBase<K, V> implements WritableKVState<K, V> {
     /** A map of all modified values buffered in this mutable state */
-    private final Map<K, V> modifications = new LinkedHashMap<>();
+    private final Map<K, V> modifications;
 
     /**
      * Create a new StateBase.
      *
      * @param stateKey The state key. Cannot be null.
      */
-    protected WritableKVStateBase(@NonNull final String stateKey) {
+    // @todo("10153") make comparator nonnull and make modifications a SortedMap
+    protected WritableKVStateBase(@NonNull final String stateKey, @Nullable final Comparator<K> comparator) {
         super(stateKey);
+        if (comparator != null) modifications = new TreeMap<>(comparator);
+        else modifications = new LinkedHashMap<>();
```

Comment:
I think we might have missed on additional case here, but it isn't clear how to add it in (darn Erasure behaviors!).
Basically, if `K` is a `Comparable`, then we should use the TreeMap in that case as well, without an explicit comparator.  We don't have an instance of `K` to test here, however, so I'm unsure how to test the type (or if we even could).
Probably not worth worrying about at this time, but something perhaps to note as a future issue.

Reply:
Agree we should use TreeMap when K is Comparable. Generics Type Erasure is hard to check here. We will make comparator @NonNull when all services have implemented their comparator key. That will cover all cases. We should stick to either Comparator or Comparable, not mix.